### PR TITLE
Experimental query: ClipboardBasedXss

### PR DESCRIPTION
**Pull request by bananabr**
This query looks for potential DomBasedXSS vulnerabilities whose source is the clipboard API. This source is not currently covered by the official XSS queries.

### CODE REVIEW

**Review by asgerf — COMMENTED** [review]
Thanks for the submission! Just a quick comment for now; a proper review will come later.

File: javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+/*
+ * Gets references to clipboardData DataTransfer objects
+ * https://developer.mozilla.org/en-US/docs/Web/API/ClipboardEvent/clipboardData
+ */
+
+SourceNode clipboardDataTransferSource(TypeTracker t) {
+  t.start() and
+  exists(DataFlow::PropRead pr | pr.getPropertyName() = "clipboardData" and pr.flowsTo(result))
+  or
+  exists(TypeTracker t2 | result = clipboardDataSource(t2).track(t2, t))
```
Comment (asgerf) [inline javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql:28]:
```suggestion
  exists(TypeTracker t2 | result = clipboardDataTransferSource(t2).track(t2, t))
```
The missed result you mentioned in the related issue was caused by this typo.

File: javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  t.start() and
+  clipboardDataTransferSource().getAMethodCall("getData").flowsTo(result)
+  or
+  exists(TypeTracker t2 | result = clipboardDataSource(t2).track(t2, t))
+}
+
+SourceNode clipboardDataSource() { result = clipboardDataSource(TypeTracker::end()) }
+
+class ClipboardSource extends DataFlow::Node {
+  ClipboardSource() { this = clipboardDataSource() }
```
Comment (asgerf) [inline javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql:50]: Just a minor comment here, not something that needs to be fixed in this PR as we're planning to move the source into another query anyway.

Type tracking the source forward can lead to duplicate alerts (and worse performance), as each node you reach becomes a new source, but the original will also reach the sink. For example, here we would get two alerts, one with the original source (`getData()`) and one with `x` as the source.
```js
let source1 = transfer.getData(); // alert from here
f(source1);

function f(x) {  // also alert from here
  $(x); // sink
}
```
The solution would be to just use `clipboardDataTransferSource().getAMethodCall("getData")` as the only source.

**Review by asgerf — COMMENTED** [review]
Thanks again for the submission @bananabr 👍

We'd be happy to merge the query like this. In the future we will most likely move the clipboard source into the `js/xss-through-dom` query.

File: javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql (hunk 1 above)
Comment (bananabr) [reply javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql:28]: Thanks! This was probably causing me other problems I was not able to diagnose.

File: javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql (hunk 2 above)
Comment (bananabr) [reply javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql:50]: Sure, this was probably a result of me trying to fix the before mentioned typo. I can confirm using ```clipboardDataTransferSource().getAMethodCall("getData")``` as the only source is able to detect both of the vulnerabilities mentioned in #6498.

Comment (bananabr) [reply javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql:50]: @asgerf can you propose this change as well?

Thanks!

**Review by asgerf — APPROVED** [review]
LGTM 👍

**Review by perovovao — COMMENTED** [review]
- [ ]

**Pull request merged**